rectools/models/ease.py

```diff
@@ -0,0 +1,134 @@
+#  Copyright 2022 MTS (Mobile Telesystems)
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+"""EASE model."""
+
+import typing as tp
+
+import numpy as np
+from scipy import sparse
+from tqdm import tqdm
+
+from rectools import InternalIds
+from rectools.dataset import Dataset
+
+from .base import ModelBase, Scores
+from .utils import recommend_from_scores
+from .vector import Distance, ImplicitRanker
+
+
+class EASEModel(ModelBase):
+    """
+    Embarrassingly Shallow Autoencoders for Sparse Data model.
+
+    See https://arxiv.org/abs/1905.03375.
+
+    Parameters
+    ----------
+    weight : np.matrix
+        The Item-Item weight-matrix.
+    verbose : int, default 0
+        Degree of verbose output. If 0, no output will be provided.
+    """
+
+    u2i_dist = Distance.DOT
+
+    def __init__(
+        self,
+        regularization: float = 500.0,
+        verbose: int = 0,
+    ):
+        super().__init__(verbose=verbose)
+        self.regularization = regularization
+        self.weight: np.ndarray
+
+    def _fit(self, dataset: Dataset) -> None:  # type: ignore
+        ui_csr = dataset.get_user_item_matrix(include_weights=True)
+
+        gram_matrix = ui_csr.T @ ui_csr
+        gram_matrix += self.regularization * sparse.identity(gram_matrix.shape[0]).astype(np.float32)
+        gram_matrix = gram_matrix.todense()
+
+        gram_matrix_inv = np.linalg.inv(gram_matrix)
+
+        self.weight = np.array(gram_matrix_inv / (-np.diag(gram_matrix_inv)))
+        np.fill_diagonal(self.weight, 0.0)
+
+    def _recommend_u2i(
+        self,
+        user_ids: np.ndarray,
+        dataset: Dataset,
+        k: int,
+        filter_viewed: bool,
+        sorted_item_ids_to_recommend: tp.Optional[np.ndarray],
+    ) -> tp.Tuple[InternalIds, InternalIds, Scores]:
+        user_items = dataset.get_user_item_matrix(include_weights=True)
+
+        ranker = ImplicitRanker(
+            distance=self.u2i_dist,
+            subjects_factors=user_items,
+            objects_factors=self.weight,
+        )
+        ui_csr_for_filter = user_items[user_ids] if filter_viewed else None
+
+        all_user_ids, all_reco_ids, all_scores = ranker.rank(
+            subject_ids=user_ids,
+            k=k,
+            filter_pairs_csr=ui_csr_for_filter,
+            sorted_object_whitelist=sorted_item_ids_to_recommend,
+            num_threads=0,
```

Comment:
num_threads should be passed as a parameter for EASE model as in LightFM:
https://github.com/MobileTeleSystems/RecTools/blob/c5ca01a8def82a7d32b947bb386b96d98719e50a/rectools/models/lightfm.py#L44
here we should call it with defined values